### PR TITLE
Add an option to disable the rate limit and not apply the limit to admin

### PR DESCRIPTION
**Pull request by leehuwuj**

### CODE REVIEW

File: src/ragapp/backend/middlewares/rate_limit.py [hunk 1]
```diff
@@ -2,29 +2,49 @@
...
-    time_frame = _get_time_frame()
-    user_id = _extract_user_id_from_request(request)
-    request_count = UserChatService.get_user_chat_request_count(user_id, time_frame)
-    if request_count >= CHAT_REQUEST_LIMIT_THRESHOLD:
-        raise HTTPException(
-            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
-            detail=f"You have exceeded {CHAT_REQUEST_LIMIT_THRESHOLD} chat requests. Please try again later.",
+    if CHAT_REQUEST_LIMIT_ENABLED:
+        time_frame = _get_time_frame()
+        user = _extract_user_info_from_request(request)
```
Comment (marcusschiesser) [inline src/ragapp/backend/middlewares/rate_limit.py:34]: good idea with `UserInfo` - but it's not a concern of the rate limiting, please move it to a own file (it's a model) and use it here like:

```suggestion
        user = UserInfo.extract_from_request(request)
```

also move the `extract_from_request` function below then

File: src/ragapp/backend/middlewares/rate_limit.py [hunk 2]
```diff
@@ -35,7 +55,7 @@ def _get_time_frame():
     return time.strftime("%Y-%m-%d")
 
 
-def _extract_user_id_from_request(request: Request) -> str:
+def _extract_user_info_from_request(request: Request) -> UserInfo:
```
Comment (marcusschiesser) [inline src/ragapp/backend/middlewares/rate_limit.py:58]: once moved to UserInfo, how about splitting this into two functions, separating the concerns: retrieving JWT token from cookie and parsing UserInfo from JWT token

File: src/ragapp/backend/middlewares/rate_limit.py (hunk 1 above)
Comment (marcusschiesser) [reply src/ragapp/backend/middlewares/rate_limit.py:34]: the code will be way cleaner as a lot of constants will also end up in the new file

File: src/ragapp/backend/middlewares/rate_limit.py (hunk 2 above)
Comment (marcusschiesser) [reply src/ragapp/backend/middlewares/rate_limit.py:58]: also i just realized that we don't validate the JWT token - it's for a later PR, I added an issue: https://github.com/ragapp/ragapp/issues/145

Comment (leehuwuj) [reply src/ragapp/backend/middlewares/rate_limit.py]: Got it! I focused on getting the rate limit set up first, and I’ll add JWT validation later.

File: src/ragapp/backend/middlewares/rate_limit.py (hunk 1 above)
Comment (leehuwuj) [reply src/ragapp/backend/middlewares/rate_limit.py]: Good idea! But i think this is better:
```python
jwt = JWT(request.cookies)
user = UserInfo.from_jwt_data(jwt.data)
```
The thing is we may end up with other jwt data in middleware so it's good to keep the jwt here and use it to construct user info

File: src/ragapp/backend/middlewares/rate_limit.py [hunk 3]
```diff
@@ -1,37 +1,22 @@
...
-
-    @computed_field
-    def is_admin(self) -> bool:
-        return ADMIN_ROLE in self.roles
 
 
 async def request_limit_middleware(request: Request) -> Response:
     if CHAT_REQUEST_LIMIT_ENABLED:
+        jwt = JWT(request.cookies)
+        user = UserInfo.from_jwt_data(jwt.data)
```
Comment (marcusschiesser) [inline src/ragapp/backend/middlewares/rate_limit.py:18]: how about adding a convient function?
```suggestion
        user = UserInfo.from_request(request)
```

File: src/ragapp/backend/middlewares/rate_limit.py [hunk 4]
```diff
@@ -1,37 +1,22 @@
...
         time_frame = _get_time_frame()
-        user = _extract_user_info_from_request(request)
         # Use user name as the key for rate limiting
         request_count = UserChatService.get_user_chat_request_count(
             user.username, time_frame
```
Comment (marcusschiesser) [inline src/ragapp/backend/middlewares/rate_limit.py:22]: probably better to pass the user model now 
```suggestion
        request_count = UserChatService.get_user_chat_request_count(
            user, time_frame
```

File: src/ragapp/backend/models/user_info.py [hunk 5]
```diff
@@ -1,14 +1,16 @@
 from pydantic import BaseModel, computed_field
 
+from .jwt import JWT
+
 ADMIN_ROLE = "admin-manager"
-JWT_USER_ID_CLAIM = "preferred_username"  # The claim in the JWT token that stores the user ID or user name
+JWT_USER_NAME_CLAIM = "preferred_username"  # The claim in the JWT token that stores the user ID or user name
```
Comment (marcusschiesser) [inline src/ragapp/backend/models/user_info.py:6]: how about using `user_id` and using here
```suggestion
JWT_USER_ID_CLAIM = os.getenv("JWT_USER_ID_CLAIM", "preferred_username")  # The claim in the JWT token that stores the user ID (defaults to `preferred_username` from Keycloak)
```

File: src/ragapp/backend/models/user_info.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+JWT_USER_ID_CLAIM = os.getenv(
+    "JWT_USER_ID_CLAIM", "preferred_username"
+)  # The claim in the JWT token that stores the user ID (defaults to `preferred_username` from Keycloak)
+JWT_USER_ROLES_CLAIM = (
+    "X-Forwarded-Roles"  # The claim in the JWT token that stores the user roles
+)
+
+
+class UserInfo(BaseModel):
+    user_name: str
```
Comment (marcusschiesser) [inline src/ragapp/backend/models/user_info.py:17]:
```suggestion
    id: str
```

**Review by marcusschiesser — APPROVED** [review]

**Pull request merged**